docs/concepts/native-interop.rst

```diff
@@ -0,0 +1,254 @@
+two namespaces will allow you access to the attributes
+
+Let's start from the most common example, and that is reusing unmanaged functions 
+in your managed code. Let's show a message box from a command-line application:
+
+.. code-block:: c#
+    :linenos:
+    
+    using System.Runtime.InteropServices;
+
+    public class Program {
+    
+        [DllImport("user32.dll")]
+        public static extrn int MessageBox(IntPtr hWnd, String text, String caption, int options);
+
+        public static void Main(string[] args) {
+            MessageBox(IntPtr.Zero, "Command-line message box", "Attention!", 0);
+        }
+    }
+
+The example above is pretty simple, but it does show off what is needed to 
+invoke unmanaged functions from managed code. Let's step through the example:
+
+* Line #1 shows the using statement for the ``System.Runtime.InteropServices`` 
+  which is the namespace that holds all of the items we need. 
+* Line #5 introduces the ``DllImport`` attribute. This attribute is crucial, as 
+  it tells the runtime that it should load the unmanaged DLL. This is the DLL 
+  into which we wish to invoke.
+* Line #6 is the crux of the P/Invoke work. It defines a managed method that has
+  the **exact same signature** as the unmanaged one. The declaration has a new 
+  keyword that you can notice, ``extern``, which tells the runtime this is an 
+  external method, and that when you invoke it, the runtime should find it in the 
+  DLL specified in ``DllImport`` attribute.
+
+The rest of the example is just invoking the method as you would any other managed 
+method. 
+
+Of course, the runtime allows communication to flow both ways 
+which enables you to call into managed artifacts from native functions, using 
+function pointers. The closest thing to a function pointer in managed code is a 
+**delegate**, so this is what is used to allow callbacks from native code into 
+managed code. 
+
+The way to use this feature is similar to managed to native process described 
+above. For a given callback, you define a delegate that matches the signature, 
+and pass that into the external method. The runtime will take care of everything 
+else.
+
+.. code-block:: c#
+    :linenos:
+
+    using System;
+    using System.Runtime.InteropServices;
+
+    namespace ConsoleApplication1 {
+
+        class Program {
+
+            delegate bool EnumWC(IntPtr hwnd, IntPtr lParam);
+
+            [DllImport("user32.dll")]
+            static extern int EnumWindows(EnumWC hWnd, IntPtr lParam);
+
+            static bool OutputWindow(IntPtr hwnd, IntPtr lParam) {
+                Console.WriteLine(hwnd.ToInt64());
+                return true;
+            }
+
+            static void Main(string[] args) {
+                EnumWindows(OutputWindow, IntPtr.Zero);
+            }
+        }
+    }
+    
+
+Before we walk through our example, it is good to go over the signatures of the 
```

Comment:
OK, I have to ask this: why? Seems to me these are two different styles that are each pretty good, especially since you are reading this outsode of the code editor, in an environment which is mostly suited for text, not code.